metals/src/main/scala/scala/meta/internal/metals/BatchedFunction.scala

```diff
@@ -97,22 +122,25 @@ final class BatchedFunction[A, B](
       callback: () => Unit,
   )
 
-  private val lock = new AtomicBoolean()
+  private val lock = new AtomicInteger(0)
+
   private def unlock(): Unit = {
-    lock.set(false)
+    lock.set(0)
+    currentCancelable.set(None)
     if (!queue.isEmpty) {
       runAcquire()
     }
   }
   private def runAcquire(): Unit = {
-    if (!isPaused.get() && lock.compareAndSet(false, true)) {
-      runRelease()
+    val id = rand.nextInt()
```

Comment:
Random could be zero,no? This should probably be a raising number. Why do we need the additional int though?

Reply:
Before `cancel()` would clear the queue and do `unlock()`. But we still may get a response from the request that was currently processed, which would also do `ulock()` as callback. We are loosing control over the lock a bit. Not sure if that is a real problem though. Now when we also do cancel on the currently processed request, we have no guarantee the futures underneath will actually react to `cancel`, so I'm making sure that only the non-cancelled currently processed request will have `onComplete` executed. There is most probably a better way to do this.